File: 1306jumpGameIII/jumpGameIII.js
/**
 * @param {number[]} arr
 * @param {number} start
 * @return {boolean}
 */
var canReach = function(arr, start) {
    let queue = [start]
    let seen = new Set(queue)
    while (queue.length)
    {
        index = queue.shift()
        if (!arr[index]) return true
        else
        {
            const child = new Set([index + arr[index], index - arr[index]].filter(_ => 0 <= _ && _ < arr.length))
            seen.forEach(_ => child.delete(_))
            queue.push(...child)
            child.forEach(_ => seen.add(_))
        }
    }
    return false
};

File: 1306jumpGameIII/jumpGameIII.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        final LinkedList<Integer> queue = IntStream.of(start).boxed().collect(Collectors.toCollection(LinkedList::new));
        final Set<Integer> seen = IntStream.of(start).boxed().collect(Collectors.toSet());
        while (!queue.isEmpty())
        {
            final int index = queue.poll();
            if (arr[index] == 0) return true;
            else
            {
                final Set<Integer> child = IntStream.of(index + arr[index], index - arr[index]).filter($ -> 0 <= $ && $ < arr.length).boxed().collect(Collectors.toSet());
                child.removeAll(seen);
                queue.addAll(child);
                seen.addAll(child);
            }
        }
        return false;
    }
}
